### ISSUE TITLE
The inventory information for utility items is incorrect.

### ISSUE THREAD

**Issue by JuhaKiili** [markus-wa/demoinfocs-golang#459]
**Describe the bug**
The inventory is not accurately synchronized with the actions of throwing, dropping, or picking up projectiles.

**To Reproduce**
This issue appears to be prevalent in most demos. However, I will provide two specific examples for reference:

eg-gold-vs-hyperion-m2-ancient.dem
https://www.hltv.org/matches/2367280/eg-gold-vs-hyperion-esl-impact-league-season-4-north-america 

spirit-vs-9ine-m2-mirage.dem
https://www.hltv.org/matches/2367004/spirit-vs-9ine-cct-north-europe-series-8

To analyze and identify the issue, I have utilized a test script that outputs information on inventory status as well as instances of picks, drops, and throws.

Please note that the mentioned demos and links serve as examples, and the issue is observable in various other instances as well.

```
go run debug_script.go -demo spirit-vs-9ine-m2-mirage.dem
```

debug_script.go
```go
package main

import (
	"fmt"
	"os"
	"sync"

	ex "github.com/markus-wa/demoinfocs-golang/v4/examples"
	demoinfocs "github.com/markus-wa/demoinfocs-golang/v4/pkg/demoinfocs"
	common "github.com/markus-wa/demoinfocs-golang/v4/pkg/demoinfocs/common"
	events "github.com/markus-wa/demoinfocs-golang/v4/pkg/demoinfocs/events"
)

func main() {
	f, err := os.Open(ex.DemoPathFromArgs())
	checkError(err)

	defer f.Close()

	p := demoinfocs.NewParser(f)
	defer p.Close()

	// Parse header
	header, err := p.ParseHeader()
	checkError(err)
	fmt.Println("Map:", header.MapName)

	var grenadeThrowers sync.Map
	var freezetime bool
	var roundNum int

	// Register round start
	p.RegisterEventHandler(func(e events.RoundStart) {
		roundNum += 1
		fmt.Printf("\n============ ROUND START %d ============\n", roundNum)
		freezetime = true
	})

	// Register freezetime end
	p.RegisterEventHandler(func(e events.RoundFreezetimeEnd) {
		fmt.Println("\n--- inventory after freezetime ---\n")
		// Print all grenades from all players
		for _, player := range p.GameState().Participants().Playing() {
			printGrenadeInventory(player)
		}
		fmt.Println("\n--- Let's go! ---\n")
		freezetime = false
	})

	// Register dropping grenades after freezetime
	p.RegisterEventHandler(func(e events.ItemDrop) {
		if e.Weapon.Type.Class() == common.EqClassGrenade && !freezetime {
			fmt.Printf("%s dropped a %s\n\n", e.Player.Name, e.Weapon.Type)
		}
	})

	// Register picking up grenades after freezetime
	p.RegisterEventHandler(func(e events.ItemPickup) {
		if e.Weapon.Type.Class() == common.EqClassGrenade && !freezetime {
			fmt.Printf("%s picked up a %s\n\n", e.Player.Name, e.Weapon.Type)
		}
	})

	// Register handler on weapon fire for grenades
	p.RegisterEventHandler(func(e events.WeaponFire) {
		if e.Weapon.Type.Class() == common.EqClassGrenade {
			fmt.Printf("%s threw a %s\n", e.Shooter.Name, e.Weapon.Type)
			grenadeThrowers.Store(e.Shooter, struct{}{})
		}
	})

	// Register frame done
	p.RegisterEventHandler(func(e events.FrameDone) {
		hadThrowers := false

		grenadeThrowers.Range(func(key, value interface{}) bool {
			thrower := key.(*common.Player)
			printGrenadeInventory(thrower)

			hadThrowers = true
			return true
		})

		// Clear the map after processing
		grenadeThrowers = sync.Map{}
	})

	// Register round end
	p.RegisterEventHandler(func(e events.RoundEnd) {
		fmt.Println("--- Round end inventory ---\n")
		// Print all grenades from all players
		for _, player := range p.GameState().Participants().Playing() {
			printGrenadeInventory(player)
		}
	})

	// Parse to end
	err = p.ParseToEnd()
	checkError(err)
}

func printGrenadeInventory(player *common.Player) {
	weapons := player.Weapons()
	if weapons == nil || len(weapons) <= 0 {
		fmt.Printf("%s has no weapons\n", player.Name)
	} else {
		fmt.Printf("%s inventory: ", player.Name)
		for _, weapon := range weapons {
			if weapon != nil && weapon.Type.Class() == common.EqClassGrenade {
				if weapon.AmmoInMagazine() + weapon.AmmoReserve() != 1 {
					fmt.Printf("%s (%d), ", weapon.String(), weapon.AmmoInMagazine() + weapon.AmmoReserve())
				} else {
					fmt.Printf("%s, ", weapon.String())
				}
			}
		}
		fmt.Println()
	}
}

func checkError(err error) {
	if err != nil {
		panic(err)
	}
}
```

**Current behavior**

eg-gold-vs-hyperion-m2-ancient.dem
```
============ ROUND START 5 ============

--- inventory after freezetime ---

JULI inventory: Smoke Grenade, HE Grenade, 

--- Let's go! ---

JULI threw a HE Grenade
JULI inventory: HE Grenade, Smoke Grenade, 

JULI threw a Molotov
JULI inventory: Smoke Grenade, HE Grenade, 

--- Round end inventory ---

JULI inventory: Smoke Grenade (2), HE Grenade, 

```

**Expected behavior:**
- The molotov should be present in the inventory before it is thrown. Currently, it seems to appear out of nowhere.
- The HE grenade should not be in the inventory after it has been thrown.
- The player should not have two smoke grenades after initially starting with only one.

**Current behavior**

spirit-vs-9ine-m2-mirage.dem
```
============ ROUND START 2 ============

--- inventory after freezetime ---

Kylar inventory: Smoke Grenade, Flashbang, Incendiary Grenade, HE Grenade, 

--- Let's go! ---

Kylar threw a Incendiary Grenade
Kylar inventory: Incendiary Grenade, HE Grenade, Smoke Grenade, Flashbang, 

Kylar threw a Smoke Grenade
Kylar inventory: Flashbang, 

--- Round end inventory ---

Kylar inventory: Flashbang, 
```

**Expected behavior:**
- The HE grenade should remain in the inventory during and after the round since it was not thrown.

**Library version**
aa1b9255d3d72d9b0ab497c87ecb6d604562c4f6

**Comment by markus-wa** [comment]
@JuhaKiili would you be able to check if #477 fixes the issue?

**Comment by markus-wa** [comment]
fixed in v4.0.0
